### DISCUSSION TITLE
startsWith(BaseUrl) in redirect callback is missing "/" at the end of URL

### DISCUSSION THREAD

**Discussion by jjojo** [nextauthjs/next-auth#3948] [Ideas]
### Description 🐜

First off, thanks for a really nice lib! 🙏 

@iaincollins explains very well in https://github.com/nextauthjs/next-auth/issues/591#issuecomment-699906051 that the baseUrl is checked to prevent malicious sites (clones) to steal credentials from your users. However, since baseUrl dont include an ending "/" even though one is provided to `NEXTAUTH_URL=http://localhost:3000/` like so. This would make it possible to redirect to for example `http://localhost:30001` on the localhost example. But in production if `https://example.com` is the baseUrl that makes `https://example.companyevilcredentialstealers.com` a valid redirect as well since it checks with `"https://example.companyevilcredentialstealers.com".startsWith("https://example.com")`, and that is true.

I'm sure somewhere the urls are parsed/created with new URL(). If so, using the `href` should always return the full url with the ending "/"

My simple workaround for now looks like this:
```js
  redirect: async ({url, baseUrl: noSlash}) => {
      const baseUrl = new URL(noSlash).href; // basically I just added this and it seems to do the trick
      return url.startsWith(baseUrl)
        ? Promise.resolve(url)
        : Promise.resolve(baseUrl);
  },
``` 

### Is this a bug in your own project?

No

### How to reproduce ☕️

Just go to https://next-auth-example.vercel.app and try for yourself. 
Entering a callback url like this for example 
```
https://next-auth-example.vercel.app/api/auth/signin?callbackUrl=https://next-auth-example.vercel.app.evilcredentialstealers.org
```
Will redirect you to 
```
https://next-auth-example.vercel.app.evilcredentialstealers.org
```
This is not an existing domain of course. But it could be if you owned the evilcredentialstealers.org domain and registered `next-auth-example.vercel.app` as a subdomain.

### Environment 🖥

 **System**:
   [Irrelevant]
  **Binaries**:
    Node: 14.18.1 - ~/.nvm/versions/node/v14.18.1/bin/node
    npm: 6.14.15 - ~/.nvm/versions/node/v14.18.1/bin/npm
  **Browser**:
    Chrome: 96.0.4664.45
  **npmPackages**:
    next: 11.1.2 => 11.1.2
    next-auth: ^4.0.0-beta.6 => 4.0.0-beta.6
    react: 17.0.2 => 17.0.2

### Contributing 🙌🏽

Yes,  I am willing to help solve this bug in a PR

**Comment by balazsorban44** [comment]
The flow is as follows:

1.
https://github.com/nextauthjs/next-auth/blob/0f132de1156a841d6bd0b0462c2947c1d44846fb/src/core/init.ts#L43

2.
https://github.com/nextauthjs/next-auth/blob/beta/src/lib/parse-url.ts

3.
https://github.com/nextauthjs/next-auth/blob/0f132de1156a841d6bd0b0462c2947c1d44846fb/src/core/lib/callback-url.ts#L25-L34

4.
https://github.com/nextauthjs/next-auth/blob/0f132de1156a841d6bd0b0462c2947c1d44846fb/src/core/lib/default-callbacks.ts#L7-L11

**Comment by jjojo** [comment]
Thanks for the references 🙏 
I think the issue originates from parseUrl and `url.origin` since that wont include an ending slash. It might be an oversimplification to just change to `url.href` though now that I think about it. I'll try to find some time this weekend and open a PR if I find a good solution to this :)
